features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,307 @@
+additional information can be included. If any of the delievery responses
+disallowed the operation, the operation is considered as failed, with the
+disallowing reasons and additional information as part of the error. For example:
+```json
+{
+    "error": {
+        "name": "WebHookError",
+        "code": 10000,
+        "message": "Operation is disallowed by web-hook",
+        "info": {
+            "errors": [
+                {
+                    "reason": "the metadata does not match the required format.",
+                    "data": {
+                        "email": "invalid email format"
+                    }
+                }
+            ]
         }
     }
 }
 ```
 
-Note that, `context` is above format is still under discussion.
+The total time spent in all deliveries of the event must not exceed 10 seconds,
+otherwise the operation would be considered failed.
 
-## Responding to a webhook
+BEFORE events would not be persisted and their failed deliveries would not be
+retried.
 
-To acknowledge receipt of a webhook, webhook endpoint should return a 2xx HTTP status code. All response codes outside this range, including 3xx codes, will indicate to skygear the request is failed.
+The operation is considered as failed if any of the deliveries failed. A failed
+operation would not trigger AFTER events.
 
-For some events, it may allow to webhook to modify changes (e.x. auth gear's before family hooks) when response code is 2xx. The behavior is defined by each gear, so will not be covered in this spec. Missing field in the return object means unmodified, so empty body is accepted. Another thing is gear will validate return object, if validation failed, gear should return an error to SDK.
+### AFTER Events
 
-Take auth gear as an example, its before family hooks accept that a hook returns a user object to modify current user object.
+AFTER events would be delivered to web-hook handlers asynchronously after the
+operation is performed (i.e. commited into database).
 
-```javascript=
-var express = require('express');
-var bodyParser = require('body-parser');
+All AFTER events, regardless whether handler is present, are presisted into
+database, with minimum retention period of 30 days.
 
-var app = express();
+The response body of web-hook handler is ignored.
 
-app.use(bodyParser.json());
+If any delivery failed, all deliveries will be retried after some time,
+regardless whether some deliveries may succeed. The retry will be performed
+with a variant of exponential back-off algorithm. If `Retry-After` HTTP header
+is present in the response, the delivery will not be retried before the
+specified time.
 
-app.post('/before_signup', function(req, res){
-    var user = req.body.data.user;
+If the delivery continue to fail after 3 days from the time of first attempted
+delivery, the event is marked as permanently failed and will not be retried
+automatically.
 
-    user.metadata.loveCat = true;
+## Mutations
```

Comment:
I notice different handlers have different mutation now.

But would it be better if we have the same set of data that users can mutate for all auth hooks?

The main reason is to just make it easier for developers, so they don't need to go back to the doc every time to understand what fields are mutable. 

On the other hand, it also doesn't seem we could imagine what are the possible use cases anyway, like god knows maybe users want to put some default metadata `before_user_create` and currently it is not mutable. It would drives us crazy to figure out all possible use cases at this stage.

Reply:
Initial metadata can be mutated in `before_user_create` event handler. At the moment, the possible mutations are based on the use cases you mentioned before: populating metadata according to SSO profile and auto-verify/disable based on identity.

I don't think we need to allow mutations of same fields for all events. Developer need to know what events are available, and what fields is contained in payload anyways.

Reply:
As mentioned above, it doesn't seem we could imagine all possible use cases.

I can simply put another story up, let's say someone want to have a metadata to count the number of times a user in auth got disabled. `metadata` is not mutatable in `user_disabled_status_update` right?

Are you going to add that back to mutations of `disabled_status_update`?



> I don't think we need to allow mutations of same fields for all events. Developer need to know what events are available, and what fields is contained in payload anyways.

Developer "need to know" != Good developer experience. It is just way more simple to have a single set of fields that we can mutate for the developer in all auth hooks. 

On the other hand, do you have any reason why you want to complicate it with different data are mutable for different hooks?

Reply:
For your story, I'd say you should not do this in before events: you cannot be sure the user will be disabled in before events. After events should be used instead, and we do not encourage user to save this in metadata anyways.

I'd say allowing mutation of the same set of fields in all events would complicate things:
- what'd happen if user is disabled in `before_session_create`?
- do we generates `user_metadata_update` events if user modify metadata in any hooks?
To put it simple, I can't ensure allowing arbitrary fields in before events would make sense, maintainable, or have no security implications. If this is allowed, we have to support it for a long time until next major version change. I'd prefer keeping the bare minimum useful mutations at the moment.

Reply:
> For your story, I'd say you should not do this in before events: you cannot be sure the user will be disabled in before events. After events should be used instead

So are you suggesting we have different mutation fields in before and after hook?

> and we do not encourage user to save this in metadata anyways.

Why we do not encourage user to save this in metadata anyways?

>     * what'd happen if user is disabled in `before_session_create`?

It seems obviously logically the session won't be created? Or what is the problem?

>     * do we generates `user_metadata_update` events if user modify metadata in any hooks?

This one is interesting, have you think about it for `user_create` hooks?



>     * To put it simple, I can't ensure allowing arbitrary fields in before events would make sense, maintainable, or have no security implications. If this is allowed, we have to support it for a long time until next major version change. I'd prefer keeping the bare minimum useful mutations at the moment.

I don't against try to keep it minimal, but could we have a subset that is small enough and common for every hooks? I think the additional mental burden for developers to check mutation is too high. If we have a minimal set of attributes that is mutable, we can still consider their implication and most hook's implication are relatively similar? (Like the metadata update event above)

Reply:
> So are you suggesting we have different mutation fields in before and after hook?
We do not support mutations requested by after event handlers: they're asynchronous.

> Why we do not encourage user to save this in metadata anyways?
We encourage developer to store application-specific data in their own database right?

> It seems obviously logically the session won't be created? Or what is the problem?
Your suggested behavior is logically sound. The problem is the behavior would be different for different events. If it is signing up, would disabling the user fail the signup, or return the user object without a session? If user is already logged in, would disabling the user failed the operation, revoke the session, or whatever?

> This one is interesting, have you think about it for user_create hooks?
This has been noted in the spec: `user_create` is not updating metadata, but creating. So `user_metadata_update` event would not be generated.

> I don't against try to keep it minimal, but could we have a subset that is small enough and common for every hooks?
I think the current list of mutation is reasonable: developer can mutate whatever the event is mutating. When user is creating, the whole user object is mutated so all reasonable fields are available. When an API is used to modify a field, that specific field would be available for mutation.

Reply:
> We do not support mutations requested by after event handlers: they're asynchronous.

Why async event handler = no mutations?

> We encourage developer to store application-specific data in their own database right?

I think the line is actually quite blurred, just like I'm not sure what you meant by application-specific data neither. Practically we would limit the functionality of metadata and users need to judge what they want to do with it, and probably there are 3 reasons they might want to put something in metadata:
1. Because it needs to be displayed in User Managemenet Dashboard or Auth UIKit that Skygear's code needs it to be accessible.
2. Because it is used as claims or something else for identity provider (the feature is not there yet)
3. Some other reason that is beneficial, e.g. we want to have built-in field-level encryption per-key in metadata

So in that case, maybe number of disable is a value they want to see in User Management Dashboard (which makes perfect sense, maybe admin user need it to decide if the user is fishy or for customer support reason)

> The problem is the behavior would be different for different events. If it is signing up, would disabling the user fail the signup, or return the user object without a session? If user is already logged in, would disabling the user failed the operation, revoke the session, or whatever?

Sounds like something we can still design for?

1. If user is already logged in and disabled, it won't affect the current operation / revoke the session, just not able to login (Unless our disable user function will revoke all sessions immediately, in that case we should be consistent and revoke the session when disabled is set true)
1. disabled during signup: this one is more tricky, if we don't revoke sessions after a disabled field is set as currently, I suggest we should just be consistent and return the session first.

> This has been noted in the spec: `user_create` is not updating metadata, but creating. So `user_metadata_update` event would not be generated.

I don't see it being specified anywhere. If you meant the words "initialize" meant "not updating", I think you need to be specific with it in the spec.

Similar, maybe we should also specify would metadata itself being updated by mutations at `user_metadata_update` webhook would not trigger another update? (or would it?)

> I think the current list of mutation is reasonable: developer can mutate whatever the event is mutating. When user is creating, the whole user object is mutated so all reasonable fields are available. When an API is used to modify a field, that specific field would be available for mutation.

You're missing the point, it would be much easier for developers to use Skygear if the mutations list is common, so I'm suggesting a minimal list of **common** mutations.

Another point is the current list doesn't seems cover enough use case. If I'm not sitting down to do some serious work at thinking about the use case, but I can still think of a lot of use cases that current design can't cater, I won't think the list is "reasonable".

Here are a random list of use cases...
* Count number of disabled happens for a user (as mentioned above)
* Manually update `verified` status when new identity is created, maybe some special rules for an app
* Disable a user automatically after a certain number of login event

So maybe we need to think about the use cases carefully, and make sure the list is reasonable.

OR, since there are not many fields in User Object actually, we can enumerate it and make a common list that make sense and should cover most use cases. (In fact I don't recall there are other fields we want to let users to mutate besides disabled, verified, metadata (or verified_info).

@louischan-oursky please help chip in.

Reply:
> Why async event handler = no mutations?

We cannot guarantee the requested mutation would be successful in async event handler. Developer should call auth gear API directly and handle possible errors.

> I don't see it being specified anywhere.

It is noted in the description of `user_metadata_update` event:
```
**NOTE**
- To handle metadata creation, use `before/after_user_create` instead.
```

> we should also specify would metadata itself being updated by mutations at user_metadata_update webhook would not trigger another update?

It would not; we would not keep generating `before_XXX_update` events until the data become stable. As stated in design document:
```
If an entity is mutated, subsequent events will used the mutated value.

Developer is responsible for ensuring correct order of event delivery. For
example, in most case, developer would like mutating event-handlers (e.g. 
populating default value for some fields) before non-mutating event-handlers
(e.g. validating field values).
```